include/shared_memory_cache_helper.cuh

```diff
@@ -11,65 +13,225 @@ namespace quda
 {
 
   /**
-     @brief Class which wraps around a shared memory cache for a
-     Vector type, where each thread in the thread block stores a
-     unique Vector in the cache which any other thread can access.
-     Presently, the expectation is that Vector is synonymous with the
-     ColorSpinor class, but we could extend this to apply to the
-     Matrix class as well.
+     @brief Class which wraps around a shared memory cache for type T,
+     where each thread in the thread block stores a unique value in
+     the cache which any other thread can access.
+
+     This accessor supports both explicit run-time block size and
+     compile-time sizing.
+
+     * For run-time block size, the constructor should be initialied
+       with the desired block size.
+
+     * For compile-time block size, no arguments should be passed to
+       the constructor, and then the second and third template
+       parameters correspond to the y and z dimensions of the block,
+       respectively.  The x dimension of the block will be set
+       according the maximum number of threads possible, given these
+       dimensions.  To prevent shared-memory bank conflicts this width
+       is optionally padded to allow for access along the y and z dimensions.
    */
-  template <typename real, typename Vector> class VectorCache
+  template <typename T, int block_size_y = 1, int block_size_z = 1, bool pad = false, bool dynamic = true>
+  class SharedMemoryCache
   {
+    /** maximum number of threads in x given the y and z block sizes */
+    static constexpr int max_block_size_x = device::max_block_size<block_size_y, block_size_z>();
+
+    /** pad in the x dimension width if requested to ensure that it isn't a multiple of the bank width */
+    static constexpr int block_size_x = !pad ? max_block_size_x :
+      ((max_block_size_x + device::shared_memory_bank_width() - 1) /
+       device::shared_memory_bank_width()) * device::shared_memory_bank_width();
+
+    //using atom_t = sizeof(T) % 16 == 0 ? int4 : sizeof(T) % 8 == 0 ? int2 : int;
+    using atom_t = int;
```

Comment:
Wonder why the large `atom_t` type are not used here?

Reply:
```using atom_t = std::conditional_t<sizeof(T) % 16 == 0, int4, std::conditional_t<sizeof(T) % 8 == 0, int2, int>>;```

Reply:
To be honest, I can't remember why.  On older architectures, greater than 32-bit accesses would lead to a stall, but I think it's not an issue on Volta and Ampere to use 64-bit and 128-bit transaction size.  Agreed?

Reply:
I think I left this as a to do, that I forgot to come back to 😉

Reply:
I don't think there is issue on Volta and Ampere with using larger transaction size. In fact I can't even think of a down side of using it. We can just add a CC macro to prevent it from messing things up on older cards?

Reply:
Yep, sounds good.  I'll do my home work to work out what is needed there, and implement as needed.

Reply:
85da62a